src/main/scala/is/hail/methods/PCA.scala

```diff
@@ -56,12 +57,66 @@ trait PCA {
     (svd.V.multiply(DenseMatrix.diag(svd.s)), optionLoadings, someIf(computeEigenvalues, svd.s.toArray.map(math.pow(_, 2))))
   }
 
-  def doubleMatrixFromVSM(vsm: VariantSampleMatrix, getVariants: Boolean): (Option[Array[Variant]], IndexedRowMatrix)
+  def doubleMatrixFromVSM(vsm: VariantSampleMatrix, getVariants: Boolean): (Option[Array[Any]], IndexedRowMatrix)
+}
+
+class ExprPCA(val expr: String) extends PCA {
+  def doubleMatrixFromVSM(vsm: VariantSampleMatrix, getVariants: Boolean): (Option[Array[Any]], IndexedRowMatrix) = {
+
+    val partitionSizes = vsm.rdd2.mapPartitions(it => Iterator.single(it.size)).collect().scanLeft(0)(_ + _)
+    assert(partitionSizes.length == vsm.rdd2.getNumPartitions + 1)
+    val pSizeBc = vsm.sparkContext.broadcast(partitionSizes)
+
+    val rowType = vsm.rowType
+    val samplesBc = vsm.sampleIdsBc
+    val sampleAnnotationsBc = vsm.sampleAnnotationsBc
+
+    val ec = EvalContext(Map(
+      "global" -> (0, vsm.globalSignature),
+      "v" -> (1, vsm.vSignature),
+      "va" -> (2, vsm.vaSignature),
+      "s" -> (3, vsm.sSignature),
+      "sa" -> (4, vsm.saSignature),
+      "g" -> (5, vsm.genotypeSignature)))
+    val f = RegressionUtils.parseExprAsDouble(expr, ec)
+    ec.set(0, vsm.globalAnnotation)
+
+    val mat = vsm.rdd2.mapPartitionsWithIndex {case (i, it) =>
+      val pStartIdx = pSizeBc.value(i)
+      var j = 0
+      val ur = new UnsafeRow(rowType)
+      it.map {rv =>
+        ur.set(rv)
+        ec.set(1, ur.get(1))
+        ec.set(2, ur.get(2))
+        val gs = ur.getAs[IndexedSeq[Any]](3)
+        val row = IndexedRow(pStartIdx + j, Vectors.dense(gs.zipWithIndex.map {case (g, k) =>
```

Comment:
use a while loop, instead of zipWithIndex. This allocates two IndexedSeq vectors, I believe! D:

Reply:
D: noooooo

Reply:
yep:
```scala

scala> val r = new IndexedSeq[Int] { def length() = 5; def apply(i: Int): Int = i }
r: IndexedSeq[Int]{def length(): Int} = (0, 1, 2, 3, 4)

scala> r.zipWithIndex
res0: IndexedSeq[(Int, Int)] = Vector((0,0), (1,1), (2,2), (3,3), (4,4))

scala> r.zipWithIndex.map(_._1)
res1: IndexedSeq[Int] = Vector(0, 1, 2, 3, 4)
```